File: scala/bob/src/main/scala/bob.scala
class Bob {
  def hey(query: String) : String = {
    if(isSilent(query)) "Fine. Be that way!"
    else if(isYelling(query)) "Whoa, chill out!"
    else if(isQuestion(query)) "Sure."
    else return "Whatever."
  }

  private def isSilent(str: String) : Boolean =
    str.trim == ""

  private def isYelling(str: String) : Boolean =
    isUpperCase(str) && hasLetter(str)

  private def isUpperCase(str: String) : Boolean =
    str.toUpperCase == str

  private def hasLetter(str: String) : Boolean =
    str.exists(_.isLetter)

  private def isQuestion(str: String) : Boolean =
    str.endsWith("?")
}

File: go/bob/bob.go
// Package bob imitates a lackadaisical teenager
package bob

import (
	"strings"
)

// Hey responds as bob would
func Hey(remark string) string {
	s := strings.TrimSpace(remark)

	if isSilent(s) {
		return "Fine. Be that way!"
	}
	q := isQuestion(s)
	y := isYelling(s)

	if q && y {
		return "Calm down, I know what I'm doing!"
	} else if q {
		return "Sure."
	} else if y {
		return "Whoa, chill out!"
	}
	return "Whatever."
}

func isSilent(s string) bool {
	return len(s) == 0
}

func isYelling(s string) bool {
	return s == strings.ToUpper(s) && s != strings.ToLower(s)
}

func isQuestion(s string) bool {
	lastC := s[len(s)-1:]
	return lastC == "?"
}
